### PR TITLE
8262504: Some CLHSDB command cannot know they run on remote debugger

### PR DESCRIPTION
**Pull request by YaSuenag**
`pmap` and `pstack` CLHSDB command do not work on remote debugger, we can see following error message:

```
hsdb> pmap
not yet implemented (debugger does not support CDebugger)!
```

However, SA has different message for this purpose:

https://github.com/openjdk/jdk/blob/03d888f463c0a6e3fee70ed8ad606fc0a3082636/src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/PMap.java#L74-L78

SA should show "remote configuration is not yet implemented" when it works on remote debugger.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8262504](https://bugs.openjdk.java.net/browse/JDK-8262504): Some CLHSDB command cannot know they run on remote debugger

### Reviewers
 * [Chris Plummer](https://openjdk.java.net/census#cjplummer) (@plummercj - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2766/head:pull/2766`
`$ git checkout pull/2766`

### CODE REVIEW

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/Tool.java [hunk 1]
```diff
@@ -51,6 +51,22 @@ public Tool(JVMDebugger d) {
...
+          jvmDebugger = agent.getDebugger();
+          debugeeType = switch (agent.getStartupMode()) {
+              case HotSpotAgent.PROCESS_MODE   -> DEBUGEE_PID;
+              case HotSpotAgent.CORE_FILE_MODE -> DEBUGEE_CORE;
+              case HotSpotAgent.REMOTE_MODE    -> DEBUGEE_REMOTE;
+              default -> throw new IllegalStateException("Invalid attach mode");
+          };
+      }
+   }
+
```
Comment (plummercj) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/Tool.java:69]: This section seems to be the key to your changes, but I have to admit that I don't follow how it all ties in together. Maybe you can explain the overall structure of the Tool subclasses to help explain. They seem to be overloaded with modes that make it hard to follow all the logic. For example, you've added the `Tool(HotSpotAgent agent)` constructor, and we already have `Tool()` and `Tool(JVMDebugger d)`. It's unclear to me what the use case is for each. Also, we have about 10 subclasses of Tool. Why do only PMap an PStack meed this new constructor.

BTW, looking at Tool and its subclasses has made me realize we have a hole other area of historical tool baggage, similar to what we are discussing with CLHSDB and HSDB. My guess is all these tool subclasses used to be invoked directly from the java command line. Now they are all hidden behind `jhsdb` subcommands, but still carry the old java command line support. In addition, when adding `jhsdb` and it's subcommands, it looks like there was an effort to mimic the existing Attach API commands like jmap and jinfo, so the SA versions of these commands have multiple modes which invoke different Tool subclasses. For example, `jhdsb jmap` alone can invoke the ClassLoaderStats, FinalizerInfo, ObjectHistogram, PMap, and HeapSummary tools, in addition to dumping the heap graph in HProf or GXL format. I think things would have been simpler and cleaner if each of these were made separate `jhsdb` subcommands, but it's probably too late for that. I suppose we could deprecate the existing sub-commands and add all these new  sub-commands, but that will just make things harder for us until the old commands are eventually removed, which might never happen.

Comment (YaSuenag) [reply src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/Tool.java:69]: > This section seems to be the key to your changes, but I have to admit that I don't follow how it all ties in together. Maybe you can explain the overall structure of the Tool subclasses to help explain. They seem to be overloaded with modes that make it hard to follow all the logic. For example, you've added the `Tool(HotSpotAgent agent)` constructor, and we already have `Tool()` and `Tool(JVMDebugger d)`. It's unclear to me what the use case is for each. Also, we have about 10 subclasses of Tool. Why do only `PMap` an `PStack` need this new constructor.

All of traditional serviceability commands (jstack, jmap, etc...) seem to extend `Tool` because it has a function to parse commandline args and to attach debuggee commonly.
`Tool` supports several attach mechanism: `Tool()` supports attach debuggee manually, `Tool(JVMDebugger)` supports attach debuggee via available `JVMDegbugger`. `HotSpotAgent` communicates to debugee, so `Tool` would create its instance. Thus I added `Tool(HotSpotAgent)` to use available remote debug server.

AFAICS only `PMap` and `PStack` do not support remote debug server, so I changed for them.
(I plan to support `pmap` and `pstack` when we attach remote debug server, so it is important to know debuggee type.)

> BTW, looking at Tool and its subclasses has made me realize we have a whole other area of historical tool baggage, similar to what we are discussing with CLHSDB and HSDB. My guess is all these tool subclasses used to be invoked directly from the java command line. Now they are all hidden behind `jhsdb` subcommands, but still carry the old java command line support. In addition, when adding `jhsdb` and it's subcommands, it looks like there was an effort to mimic the existing Attach API commands like jmap and jinfo, so the SA versions of these commands have multiple modes which invoke different Tool subclasses. For example, `jhdsb jmap` alone can invoke the ClassLoaderStats, FinalizerInfo, ObjectHistogram, PMap, and HeapSummary tools, in addition to dumping the heap graph in HProf or GXL format. I think things would have been simpler and cleaner if each of these were made separate `jhsdb` subcommands, but it's probably too late for that. I suppose we could deprecate the existing sub-commands and add all these new sub-commands, but that will just make things harder for us until the old commands are eventually removed, which might never happen.

I agree with you it is too late to happen.
As I commented in https://github.com/openjdk/jdk/pull/2773#issuecomment-790203478 , we can do simplify for them as possible. For example, we can remove `main()` and related methods (e.g. showing usage) from them. If we can remove commandline support from all tool classes, It might help a little to maintain.

Comment (plummercj) [reply src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/Tool.java:69]: If I understanding correctly, we have two users of Tools.java. One is `SALauncher` and the other is `CommandProcessor`. When we use `SALauncher`, we determine the `debugeeType` based on the command line arguments used. When using the `CommandProcessor`, we need access to the `HotSpotAgent` to get the `debuggeeType`, since it has stored this in its `startupMode` field . So the only reason for this new constructor is so we can set `debugeeType`, and then as you've pointed out, that makes the following existing code work correctly:

```
 if (getDebugeeType() == DEBUGEE_REMOTE) { 
     out.println("remote configuration is not yet implemented"); 
 } else { 
     out.println("not yet implemented (debugger does not support CDebugger)!"); 
 } 
```

Comment (plummercj) [reply src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/Tool.java:69]: > I agree with you it is too late to happen.
> As I commented in [#2773 (comment)](https://github.com/openjdk/jdk/pull/2773#issuecomment-790203478) , we can do simplify for them as possible. For example, we can remove `main()` and related methods (e.g. showing usage) from them. If we can remove commandline support from all tool classes, It might help a little to maintain.

Yes, I'd like to see a lot of this excess baggage go away. I also wouldn't mind at least considering re-imagining the command line interface for all these Tool subclasses, but would only do so if we could quickly deprecate what we currently have. We already have enough cases of being able to do the same thing more than one way (I think in some cases 5 ways). We don't need one more, but on the other hand, as an example, having `jhsdb jmap` support dumping the java heap just doesn't make any sense. My guess is the first `jmap` just dumped the memory map of loaded libraries, and then more and more unrelated subcommands kept being added to it, but I'm not sure of the history. The Attach API `jmap` command suffers from the same issue, although oddly is does not support the library mapping output like the SA version does. I'm not sure which version of `jmap` came first and what it initially looked like.

Comment (YaSuenag) [reply src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/Tool.java:69]: I cannot agree with restructuring sub commands and options in jhsdb for compatibility.

Since jhsdb is introduced, SA tools (jstack, jmap, and so on) has been removed the feature to use SA, they are for live process only with Attach API. So I guess we can ignore about SA interface for them. In addition, they are marked as unsupported tool in [the document](https://docs.oracle.com/en/java/javase/15/docs/specs/man/index.html).

Thus I guess we can think about jhsdb and jcmd (dcmd) interface only. If so, I think we can remove some excess baggages from them (e.g. command line support), and it is worth to work for maintenance in future.

But this PR does not aim it, it will be the next step.

**Review by plummercj — APPROVED** [review]

**Comment by YaSuenag** [comment]
PING: could you review it? I need one more reviewer, thanks!

**Review by sspitsyn — APPROVED** [review]
Hi Yasumasa,
The fix looks good to me.
Thanks,
Serguei
